File: frontend/src/types/identity_hunt.ts
/**
 * Identity Hunt UI types (#116 Phase B).
 *
 * TypeScript mirrors of ``shared/btagent_shared/types/identity_hunt.py``.
 * String-union values match the backend StrEnum values exactly so they can be
 * compared without conversion.
 *
 * The ``IdentityFindingEvidence`` interface captures the evidence dict fields
 * that the identity detectors embed in every HuntFinding they emit —
 * ``principal_id``, ``app_id``, ``cred_type``, ``distinct_asns``, etc.  These
 * are typed as optional because the evidence dict is open-ended; the page
 * renders gracefully when fields are absent.
 */

// --------------------------------------------------------------------------- //
// Enums (mirrors Python StrEnum — lowercase string literals)
// --------------------------------------------------------------------------- //

export type IdentityProvider = "okta" | "entra" | "google_workspace" | "generic";

export type IdentityEntityKind =
  | "user"
  | "service_principal"
  | "oauth_app"
  | "session"
  | "device";

export type OAuthConsentType = "admin" | "user" | "pre_authorized" | "unknown";

export type IdentityEventKind =
  | "token_issued"
  | "token_refresh"
  | "token_revoked"
  | "mfa_challenge"
  | "mfa_denied"
  | "mfa_approved"
  | "login_success"
  | "login_failure"
  | "credential_added"
  | "credential_removed"
  | "role_assigned"
  | "role_removed"
  | "federation_trust_modified"
  | "grant_created"
  | "grant_revoked"
  | "app_consent_granted";

// --------------------------------------------------------------------------- //
// Core identity entities (mirrors IdentityEntity Pydantic model)
// --------------------------------------------------------------------------- //

export interface IdentityEntity {
  id: string;
  org_id: string;
  kind: IdentityEntityKind;
  provider: IdentityProvider;
  /** Stable, provider-scoped identifier (UPN, client_id, session token hash, etc.). */
  canonical_id: string;
  display_name: string;
  first_seen: string;
  last_seen: string;
  enrichment: Record<string, unknown>;
}

// --------------------------------------------------------------------------- //
// OAuth grant graph (mirrors OAuthGrant Pydantic model)
// --------------------------------------------------------------------------- //

export interface OAuthGrant {
  id: string;
  org_id: string;
  /** Client / application ID that holds this grant (OAuth client_id). */
  app_id: string;
  app_display_name: string;
  /** User or service account UPN / object_id the grant was consented for. */
  principal_id: string;
  provider: IdentityProvider;
  /** Normalised OAuth scope strings (e.g. 'Mail.Read', 'offline_access'). */
  scopes: string[];
  consent_type: OAuthConsentType;
  granted_at: string;
  last_used: string | null;
  /** null while the grant is active; set when revoked. */
  revoked_at: string | null;
  raw: Record<string, unknown>;
}

// --------------------------------------------------------------------------- //
// Geo location (mirrors GeoLocation Pydantic model)
// --------------------------------------------------------------------------- //

export interface GeoLocation {
  country: string;
  city: string;
  latitude: number | null;
  longitude: number | null;
  /** Autonomous System Number (e.g. 'AS15169') for ASN-diversity checks. */
  asn: string;
}

// --------------------------------------------------------------------------- //
// Identity event (mirrors IdentityEvent Pydantic model)
// --------------------------------------------------------------------------- //

export interface IdentityEvent {
  id: string;
  org_id: string;
  provider: IdentityProvider;
  kind: IdentityEventKind;
  principal_id: string;
  app_id: string;
  session_id: string;
  token_id: string;
  ip_address: string;
  geo: GeoLocation;
  user_agent: string;
  timestamp: string;
  raw: Record<string, unknown>;
}

// --------------------------------------------------------------------------- //
// Evidence dict shape — embedded inside HuntFinding.evidence
// --------------------------------------------------------------------------- //

/**
 * Fields that identity detectors embed in ``HuntFinding.evidence``.
 *
 * All fields are optional because the evidence dict is open-ended; the page
 * must degrade gracefully when any field is absent.  The full set of possible
 * fields is documented in ``IdentityDetectionResult.evidence`` (Python).
 */
export interface IdentityFindingEvidence {
  /** The stable identity principal (UPN, object_id, etc.) implicated. */
  principal_id?: string;
  /** OAuth client / application ID if the detection involves an OAuth grant. */
  app_id?: string;
  /** OAuth app display name, if known. */
  app_display_name?: string;
  /** Credential type implicated (e.g. 'refresh_token', 'access_token', 'sp_secret'). */
  cred_type?: string;
  /** Number of distinct ASNs seen for the same token/session (token-replay signal). */
  distinct_asns?: number;
  /** ASN list for token-replay findings. */
  asns?: string[];
  /** IP addresses seen across ASNs. */
  ip_addresses?: string[];
  /** OAuth scopes on the grant, if applicable. */
  scopes?: string[];
  /** Consent type for OAuth grant findings. */
  consent_type?: OAuthConsentType;
  /** Identity provider for this event. */
  provider?: IdentityProvider;
  /** Number of MFA push challenges in a window (MFA fatigue signal). */
  mfa_push_count?: number;
  /** Number of MFA denials in a window. */
  mfa_deny_count?: number;
  /** Time window in seconds over which MFA events were counted. */
  window_seconds?: number;
  /** Session ID implicated (for token-replay / session-hijack detections). */
  session_id?: string;
  /** Token ID / jti implicated. */
  token_id?: string;
  /** Raw event IDs contributing to this detection. */
  event_ids?: string[];
  /** Time of first event in the detection window. */
  window_start?: string;
  /** Time of last event in the detection window. */
  window_end?: string;
  /** App was dormant for this many days before reactivation. */
  dormant_days?: number;
  /** Whether the app previously had revoked grants (dormant reactivation). */
  previously_revoked?: boolean;
  /** Additional freeform fields from the detector. */
  [key: string]: unknown;
}

// --------------------------------------------------------------------------- //
// Timeline entry (UI-only, derived from finding + evidence)
// --------------------------------------------------------------------------- //

/**
 * A single entry in the per-principal token-lifecycle timeline.
 *
 * Constructed client-side by grouping identity ``HuntFinding``s by
 * ``principal_id`` (from ``evidence``) and sorting by ``created_at``.
 */
export interface IdentityTimelineEntry {
  /** The source HuntFinding id. */
  finding_id: string;
  /** ISO timestamp (``HuntFinding.created_at`` or ``evidence.window_end``). */
  timestamp: string;
  /** Severity from the source finding. */
  severity: string;
  /** Human-readable description derived from the finding title + evidence. */
  label: string;
  /** MITRE ATT&CK technique IDs on the finding. */
  technique_ids: string[];
  /** The raw evidence dict for drilldown. */
  evidence: IdentityFindingEvidence;
  /** The finding's current state. */
  state: string;
  /** The finding's cluster id, if any. */
  cluster_id: string | null;
}

// --------------------------------------------------------------------------- //
// Per-principal summary (UI-only)
// --------------------------------------------------------------------------- //

/**
 * All identity findings grouped by ``principal_id`` from the evidence dict.
 * Used to drive the per-principal timeline and consent panel.
 */
export interface PrincipalSummary {
  /** The principal_id string from the evidence dict. */
  principal_id: string;
  /** Highest severity across all findings for this principal. */
  max_severity: string;
  /** Total number of findings for this principal. */
  finding_count: number;
  /** Chronologically sorted timeline entries. */
  timeline: IdentityTimelineEntry[];
  /** Subset of findings whose technique_ids include consent/credential techniques. */
  consent_findings: IdentityTimelineEntry[];
}

// --------------------------------------------------------------------------- //
// OAuth grant table row (UI-only, derived from evidence)
// --------------------------------------------------------------------------- //

/**
 * A row in the OAuth grant table, extracted from finding evidence fields.
 * Multiple findings for the same principal + app are deduplicated by
 * ``app_id``.
 */
export interface GrantTableRow {
  principal_id: string;
  app_id: string;
  app_display_name: string;
  scopes: string[];
  consent_type: OAuthConsentType | "unknown";
  finding_id: string;
  severity: string;
}

File: shared/btagent_shared/types/identity_hunt.py
"""Identity Hunt schemas (Phase 6 #116 — Identity Hunt Agent).

Data contracts for the connector-independent slice of the Identity Hunt Agent.
Live-connector wiring (Okta / Entra ID / Google Workspace MCP, tracked in #100)
is deferred; everything here operates on fixture data or streaming events
supplied by the eventual connectors.

Design notes:
- No heavy deps (this is in the ``shared/`` tier — no DB, MCP, or LLM imports).
- Pydantic v2, ConfigDict(extra="forbid"), lowercase StrEnum values throughout.
- Mirrors the style of :mod:`btagent_shared.types.behavioral`.
- The ``OAuthGrant`` model is the central unit for OAuth-grant-graph detections;
  ``IdentityEvent`` is the generic event feed that detectors consume.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------


class IdentityProvider(StrEnum):
    """Cloud identity provider the entity originates from.

    Mapped 1-to-1 to the connectors in connector-tier #100. The enum exists
    here so fixture tests can run without a live connector.
    """

    OKTA = "okta"
    ENTRA = "entra"
    GOOGLE_WORKSPACE = "google_workspace"
    DUO = "duo"  # Cisco Duo MFA (Tier-2 #100)
    GENERIC = "generic"  # local / on-prem / not yet classified


class IdentityEntityKind(StrEnum):
    """The type of identity subject being tracked."""

    USER = "user"
    SERVICE_PRINCIPAL = "service_principal"
    OAUTH_APP = "oauth_app"
    SESSION = "session"
    DEVICE = "device"


class OAuthConsentType(StrEnum):
    """How the OAuth grant was consented to.

    ``admin`` grants are granted by an admin on behalf of all users;
    ``user`` grants are granted per-user; ``pre_authorized`` bypasses the
    consent dialog entirely (high-risk signal for dormant-app reactivation).
    """

    ADMIN = "admin"
    USER = "user"
    PRE_AUTHORIZED = "pre_authorized"
    UNKNOWN = "unknown"


class IdentityEventKind(StrEnum):
    """The type of identity event ingested from a provider audit log."""

    TOKEN_ISSUED = "token_issued"
    TOKEN_REFRESH = "token_refresh"
    TOKEN_REVOKED = "token_revoked"
    MFA_CHALLENGE = "mfa_challenge"
    MFA_DENIED = "mfa_denied"
    MFA_APPROVED = "mfa_approved"
    LOGIN_SUCCESS = "login_success"
    LOGIN_FAILURE = "login_failure"
    CREDENTIAL_ADDED = "credential_added"
    CREDENTIAL_REMOVED = "credential_removed"
    ROLE_ASSIGNED = "role_assigned"
    ROLE_REMOVED = "role_removed"
    FEDERATION_TRUST_MODIFIED = "federation_trust_modified"
    GRANT_CREATED = "grant_created"
    GRANT_REVOKED = "grant_revoked"
    APP_CONSENT_GRANTED = "app_consent_granted"


# ---------------------------------------------------------------------------
# Core identity entities
# ---------------------------------------------------------------------------


class IdentityEntity(BaseModel):
    """A subject in the identity plane — user, service principal, OAuth app, or session.

    ``canonical_id`` is the stable string used for clustering and suppression
    (e.g. UPN for users, client_id for OAuth apps, session_id for sessions).
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., min_length=1, max_length=200)
    org_id: str = Field(..., min_length=1, max_length=200)
    kind: IdentityEntityKind
    provider: IdentityProvider
    canonical_id: str = Field(
        ...,
        min_length=1,
        max_length=512,
        description="Stable, provider-scoped identifier (UPN, client_id, session token hash, etc.).",
    )
    display_name: str = Field(default="", max_length=300)
    first_seen: datetime
    last_seen: datetime
    enrichment: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# OAuth grant graph
# ---------------------------------------------------------------------------


class OAuthGrant(BaseModel):
    """One OAuth permission grant in the principal↔app grant graph.

    Represents a scope bundle granted to an ``app_id`` (client) on behalf of
    a ``principal_id`` (user or service account). The grant graph is built
    from provider audit logs by :mod:`btagent_shared.hunt.identity`.

    Fields map to the common OAuth 2.0 / OIDC concepts that Okta, Entra,
    and Google Workspace all expose (names differ but semantics align).
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., min_length=1, max_length=200)
    org_id: str = Field(..., min_length=1, max_length=200)
    app_id: str = Field(
        ...,
        min_length=1,
        max_length=512,
        description="Client / application ID that holds this grant (OAuth client_id).",
    )
    app_display_name: str = Field(default="", max_length=300)
    principal_id: str = Field(
        ...,
        min_length=1,
        max_length=512,
        description="User or service account UPN / object_id the grant was consented for.",
    )
    provider: IdentityProvider
    scopes: list[str] = Field(
        default_factory=list,
        description="Normalised OAuth scope strings (e.g. 'Mail.Read', 'offline_access').",
    )
    consent_type: OAuthConsentType = OAuthConsentType.UNKNOWN
    granted_at: datetime
    last_used: datetime | None = None
    # ``revoked_at`` is None while the grant is active.
    revoked_at: datetime | None = None
    # Enrichment blob — provider raw payload for forensics.
    raw: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Generic identity event (detector input)
# ---------------------------------------------------------------------------


class GeoLocation(BaseModel):
    """Coarse geolocation attached to a sign-in / token event."""

    model_config = ConfigDict(extra="forbid")

    country: str = Field(default="", max_length=64)
    city: str = Field(default="", max_length=128)
    latitude: float | None = None
    longitude: float | None = None
    asn: str = Field(
        default="",
        max_length=32,
        description="Autonomous System Number (e.g. 'AS15169') for ASN-diversity checks.",
    )


class IdentityEvent(BaseModel):
    """A single event from a provider audit log.

    Detectors in :mod:`btagent_shared.hunt.identity` operate on lists of
    these to decide whether to emit a :class:`IdentityDetectionResult`.

    The ``session_id`` / ``token_id`` fields enable token-replay and
    MFA-fatigue correlations across events from the same credential.
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., min_length=1, max_length=200)
    org_id: str = Field(..., min_length=1, max_length=200)
    provider: IdentityProvider
    kind: IdentityEventKind
    principal_id: str = Field(..., min_length=1, max_length=512)
    app_id: str = Field(default="", max_length=512)
    session_id: str = Field(
        default="",
        max_length=512,
        description="Session or refresh-token ID; used for token-replay correlation.",
    )
    token_id: str = Field(
        default="",
        max_length=512,
        description="Access-token ID / jti; used for token-replay replay detection.",
    )
    ip_address: str = Field(default="", max_length=64)
    geo: GeoLocation = Field(default_factory=GeoLocation)
    user_agent: str = Field(default="", max_length=512)
    timestamp: datetime
    raw: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Detector output — maps 1:1 to a RecordFindingRequest
# ---------------------------------------------------------------------------


class IdentityDetectionResult(BaseModel):
    """Structured output of a single identity detector run.

    Each result carries enough information to build a
    :class:`btagent_shared.types.hunt_finding.RecordFindingRequest`
    (source=identity, domain=identity) without any further enrichment.
    The mapping function lives in :mod:`btagent_shared.hunt.identity`.
    """

    model_config = ConfigDict(extra="forbid")

    detection_id: str = Field(
        ...,
        min_length=1,
        max_length=200,
        description="Unique ID for this detection hit (deterministic or ULID).",
    )
    rule_id: str = Field(..., min_length=1, max_length=200)
    title: str = Field(..., min_length=1, max_length=300)
    description: str = Field(default="", max_length=8192)
    # Severity string matches btagent_shared.types.enums.Severity values.
    severity: str = Field(default="medium")
    confidence: float = Field(default=0.7, ge=0.0, le=1.0)
    technique_ids: list[str] = Field(
        default_factory=list,
        description="MITRE ATT&CK technique IDs implicated (e.g. ['T1550.001', 'T1078']).",
    )
    # Entities + observables match the hunt_finding contract.
    entity_kind: str = Field(default="user", max_length=64)
    entity_value: str = Field(default="", max_length=512)
    observable_type: str = Field(default="", max_length=64)
    observable_value: str = Field(default="", max_length=2048)
    evidence: dict[str, Any] = Field(
        default_factory=dict,
        description="Provenance — event IDs, timestamps, geo details, etc.",
    )


# ---------------------------------------------------------------------------
# Revocation proposal (#116 Phase C) — confirmed hit → revoke-playbook (HITL)
# ---------------------------------------------------------------------------


class RevocationProposalStatus(StrEnum):
    """Lifecycle of a revoke-playbook proposal attached to an investigation."""

    PROPOSED = "proposed"
    ACCEPTED = "accepted"
    REJECTED = "rejected"


class RevocationTarget(BaseModel):
    """One (principal, app) grant slated for revocation.

    Deduped by ``(provider, principal_id, app_id)`` — several findings about
    the same grant collapse into one target carrying every source finding ID.
    """

    model_config = ConfigDict(extra="forbid")

    principal_id: str = Field(..., min_length=1, max_length=512)
    app_id: str = Field(..., min_length=1, max_length=512)
    provider: IdentityProvider
    app_display_name: str = Field(default="", max_length=300)
    scopes: list[str] = Field(default_factory=list)
    source_finding_ids: list[str] = Field(default_factory=list)


class RevocationProposal(BaseModel):
    """A revoke-playbook proposal generated when identity grant findings are
    promoted to an investigation.

    The proposal is *inert data* until a senior analyst accepts it (the HITL
    gate): acceptance materialises ``playbook_spec`` as a real SOAR playbook
    whose own first step is a second ``hitl_gate`` guarding execution.
    ``playbook_spec`` is the playbook as a JSON-safe dict — the backend dumps
    it to YAML for the playbook service on accept.
    """

    model_config = ConfigDict(extra="forbid")

    targets: list[RevocationTarget]
    rationale: str = Field(default="", max_length=8192)
    playbook_name: str = Field(..., min_length=1, max_length=300)
    playbook_spec: dict[str, Any]
    status: RevocationProposalStatus = RevocationProposalStatus.PROPOSED
    # Set on accept — the materialised playbook's ID.
    playbook_id: str | None = None
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_rationale: str = Field(default="", max_length=8192)
